include/oneapi/dpl/pstl/hetero/dpcpp/unseq_backend_sycl.h

```diff
@@ -202,101 +202,141 @@ struct __init_processing
-            const _Size __items_to_process = __adjusted_n - __adjusted_global_id;
-            // Keep these statements in the same scope to allow for better memory alignment
-            typename _AccLocal::value_type __res = __unary_op(__adjusted_global_id, __acc...);
-            for (_Size __i = 1; __i < __items_to_process; ++__i)
-                __res = __binary_op(__res, __unary_op(__adjusted_global_id + __i, __acc...));
+            _Res __res = vectorized_reduction_first<_Res>(__adjusted_global_id, __acc...);
+            for (_Size __i = 1; __i < __no_vec_ops; ++__i)
+                vectorized_reduction_remainder(__adjusted_global_id + __i * __stride, __res, __acc...);
             __local_mem[__local_idx] = __res;
+            return;
         }
-    }
-
-    template <typename _NDItemId, typename _Size, typename _AccLocal, typename... _Acc>
-    void
-    nonseq_impl(const _NDItemId __item_id, const _Size __n, const _Size __global_offset, const _AccLocal& __local_mem,
-                const _Acc&... __acc) const
-    {
-        auto __local_idx = __item_id.get_local_id(0);
-        const _Size __stride = __item_id.get_local_range(0);
-
-        const _Size __adjusted_global_id =
-            __global_offset + __item_id.get_group_linear_id() * __stride * __iters_per_work_item + __local_idx;
-        const _Size __adjusted_n = __global_offset + __n;
-
-        // Coalesced load and reduce from global memory
-        if (__adjusted_global_id + __stride * __iters_per_work_item < __adjusted_n)
+        // At least one vector operation
+        constexpr _Size __vec_size_minus_one = static_cast<_Size>(_VecSize - 1);
+        if (__adjusted_global_id + __vec_size_minus_one < __adjusted_n)
         {
-            typename _AccLocal::value_type __res = __unary_op(__adjusted_global_id, __acc...);
-            _ONEDPL_PRAGMA_UNROLL
-            for (_Size __i = 1; __i < __iters_per_work_item; ++__i)
-                __res = __binary_op(__res, __unary_op(__adjusted_global_id + __stride * __i, __acc...));
+            _Res __res = vectorized_reduction_first<_Res>(__adjusted_global_id, __acc...);
+            for (_Size __i = 1; __i < __no_vec_ops; ++__i)
+            {
+                const _Size __base_idx = __adjusted_global_id + __i * __stride;
+                if (__base_idx + __vec_size_minus_one < __adjusted_n)
+                    vectorized_reduction_remainder(__base_idx, __res, __acc...);
+                else if (__base_idx < __adjusted_n)
+                {
+                    for (_Size __idx = 0; __idx < __vec_size_minus_one; ++__idx)
+                    {
+                        if (__base_idx + __idx >= __adjusted_n)
+                            break;
+                        __res = __binary_op(__res, __unary_op(__base_idx + __idx, __acc...));
+                    }
+                }
+                else
+                    break;
+            }
             __local_mem[__local_idx] = __res;
+            return;
```

Comment:
Can we remove this return and make the subsequent `if` an `else if` or is there a performance penalty otherwise? There's a few other places we do this as well.

Reply:
Done.